File: backend/python/smartbi/api/price_anomaly.py
"""Wave2 价格异常威慑引擎 — REST API.

Routes (registered under prefix="/api/smartbi" in main.py):
- GET  /gold/price-anomaly/detect  → 检测同类物料相邻采购单价异常 (latest vs trailing-N avg)
- POST /gold/price-anomaly/ack     → 录入供应商解释/确认 (威慑非处罚, 幂等)
- GET  /gold/price-anomaly/acks    → 已确认异常列表 (报警去重 + 留痕)

Tenant-scoped: factory_id from JWT (request.state) or X-Factory-Id on internal
Java→Python calls. RBAC: 绝对金额字段 (oldPrice/newPrice/trailingAvg) 对非
price-view 角色剥零; deltaPct (率) + riskLevel/direction (威慑信号) 保留可见。

防呆 (fool-proof):
- Rule 3: ack reason_code 是 dropdown (SEASONAL/MARKET_RISE/SPEC_CHANGE/OTHER), OTHER 必填备注。
- Rule 4: ack 幂等 (同一异常只确认一次, ON CONFLICT 更新解释)。
- 报警 message 含上下文 (品名 + 供应商 + 新旧价) 便于威慑追责。
"""
from __future__ import annotations

import logging
from datetime import datetime
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Body, HTTPException, Query, Request

from smartbi_compat._rbac_strip import PRICE_VIEW_ROLES, strip_price_for_role

logger = logging.getLogger(__name__)
router = APIRouter(tags=["PriceAnomaly"])

_ACK_WRITE_ROLES: frozenset[str] = frozenset({
    "factory_super_admin",
    "platform_admin",
    "permission_admin",
    "restaurant_manager",
})

# 绝对金额 camelCase keys not matched by the shared _MONEY_PATTERN. deltaPct is a
# RATE (conveys trend/deterrence without leaking the amount) so it stays visible.
_EXTRA_MONEY_KEYS: frozenset[str] = frozenset({"trailingAvg"})


def _get_factory_id(request: Request) -> Optional[str]:
    return getattr(request.state, "factory_id", None)


def _get_role(request: Request) -> Optional[str]:
    return getattr(request.state, "role", None)


def _get_username(request: Request) -> Optional[str]:
    return getattr(request.state, "username", None)


def _require_ack_write_role(request: Request) -> None:
    role = _get_role(request)
    if role not in _ACK_WRITE_ROLES:
        raise HTTPException(
            status_code=403,
            detail=(
                "无权记录价格异常解释；请由店长/老板账号处理，"
                "仓管现场涨价说明请在送货单验收入库页填写。"
            ),
        )


def _strip_extras(node: Any) -> None:
    if isinstance(node, dict):
        for k, v in list(node.items()):
            if k in _EXTRA_MONEY_KEYS and not isinstance(v, (dict, list)):
                node[k] = None
            else:
                _strip_extras(v)
    elif isinstance(node, list):
        for item in node:
            _strip_extras(item)


def _apply_rbac_strip(data: Any, role: Optional[str]) -> Any:
    """Null absolute monetary fields for roles outside PRICE_VIEW_ROLES."""
    if role and role in PRICE_VIEW_ROLES:
        return data
    if data is None:
        return data
    strip_price_for_role(data, role)
    _strip_extras(data)
    return data


def _parse_date(s: Any):
    if not s:
        return None
    if isinstance(s, str):
        try:
            return datetime.strptime(s, "%Y-%m-%d").date()
        except ValueError:
            return None
    return s


@router.get("/gold/price-anomaly/detect")
async def detect(
    request: Request,
    trailing_n: int = Query(3, ge=1, le=20, description="count 模式: trailing 均价窗口 N"),
    epsilon_pct: float = Query(5.0, ge=0.0, le=100.0, description="异常容限 ε (百分比)"),
    baseline_mode: str = Query(
        "count", pattern="^(count|days)$",
        description="基准模式: count=trailing-N 次均价 (legacy); days=N 天移动均价 (邓总锁定 90 天)"),
    window_days: int = Query(
        90, ge=1, le=365, description="days 模式: 移动均价窗口天数 (邓总锁定 90)"),
) -> Dict[str, Any]:
    """检测同类物料相邻采购单价异常。

    返回 list (HIGH risk 优先, 然后 |delta| 大优先):
      {normalizedName, ingredientName, supplierId, supplierName, unit,
       anomalyDeliveryDate, oldPrice, newPrice, trailingAvg, deltaPct,
       direction, consecutiveAnomalyCount, riskLevel}

    baseline_mode='days' + window_days=90 = 邓总锁定的"自身 90 天移动均价"基准
    (供应商涨价预警)。复用 #53 检测内核 (同 agg_supplier_price 历史 + ack/累计高风险),
    不分叉新 detector。
    """
    factory_id = _get_factory_id(request)
    if not factory_id:
        return {"success": False, "message": "missing factory context"}

    from smartbi.config import get_pg_pool
    import smartbi.gold.price_anomaly as mod
    pool = await get_pg_pool()
    if pool is None:
        return {"success": False, "message": "smartbi_db pool unavailable"}

    try:
        data = await mod.detect_price_anomalies(
            pool, factory_id, trailing_n=trailing_n, epsilon_pct=epsilon_pct,
            baseline_mode=baseline_mode, window_days=window_days,
        )
    except ValueError as e:
        return {"success": False, "message": str(e)}
    except Exception as e:
        logger.exception("[price-anomaly] detect failed for %s", factory_id)
        return {"success": False, "message": f"detect failed: {e}"}

    _apply_rbac_strip(data, _get_role(request))
    return {"success": True, "data": data}


@router.post("/gold/price-anomaly/ack")
async def ack(
    request: Request,
    body: Dict[str, Any] = Body(...),
) -> Dict[str, Any]:
    """录入供应商对一条价格异常的解释/确认 (威慑非处罚)。

    Body (camelCase): {normalizedName, ingredientName, supplierId, supplierName,
      anomalyDeliveryDate (YYYY-MM-DD), oldPrice, newPrice, deltaPct,
      reasonCode (SEASONAL|MARKET_RISE|SPEC_CHANGE|OTHER), reasonNote}

    acknowledged_by 取自 JWT username (审计留痕)。幂等 (ON CONFLICT 更新解释)。
    """
    _require_ack_write_role(request)
    ctx_factory = _get_factory_id(request)
    body_factory = body.get("factoryId")
    factory_id = ctx_factory or body_factory
    if not factory_id:
        return {"success": False, "message": "missing factory context"}
    if ctx_factory and body_factory and ctx_factory != body_factory:
        return {"success": False,
                "message": f"factory mismatch: ctx={ctx_factory} body={body_factory}"}

    from smartbi.config import get_pg_pool
    import smartbi.gold.price_anomaly as mod
    pool = await get_pg_pool()
    if pool is None:
        return {"success": False, "message": "smartbi_db pool unavailable"}

    try:
        ack_id = await mod.record_anomaly_ack(
            pool, factory_id,
            normalized_name=body.get("normalizedName"),
            ingredient_name=body.get("ingredientName"),
            supplier_id=body.get("supplierId"),
            supplier_name=body.get("supplierName"),
            anomaly_delivery_date=_parse_date(body.get("anomalyDeliveryDate")),
            old_price=body.get("oldPrice"),
            new_price=body.get("newPrice"),
            delta_pct=body.get("deltaPct"),
            reason_code=body.get("reasonCode") or "",
            reason_note=body.get("reasonNote"),
            acknowledged_by=_get_username(request),
        )
    except ValueError as e:
        return {"success": False, "message": str(e)}
    except Exception as e:
        logger.exception("[price-anomaly] ack failed for %s", factory_id)
        return {"success": False, "message": f"ack failed: {e}"}

    return {"success": True, "data": {"id": ack_id},
            "message": "解释已记录, 该异常已确认"}


@router.get("/gold/price-anomaly/acks")
async def acks(
    request: Request,
    ingredient: Optional[str] = Query(None, description="按食材归一名过滤 (可选)"),
) -> Dict[str, Any]:
    """已确认异常列表 (报警去重 + 留痕展示)。"""
    factory_id = _get_factory_id(request)
    if not factory_id:
        return {"success": False, "message": "missing factory context"}

    from smartbi.config import get_pg_pool
    import smartbi.gold.price_anomaly as mod
    pool = await get_pg_pool()
    if pool is None:
        return {"success": False, "message": "smartbi_db pool unavailable"}

    norm = None
    if ingredient:
        # mirror the detection/ETL normalization so the filter matches.
        norm = " ".join(str(ingredient).lower().strip().split())

    try:
        data: List[dict] = await mod.list_anomaly_acks(pool, factory_id, normalized_name=norm)
    except ValueError as e:
        return {"success": False, "message": str(e)}
    except Exception as e:
        logger.exception("[price-anomaly] acks list failed for %s", factory_id)
        return {"success": False, "message": f"query failed: {e}"}

    _apply_rbac_strip(data, _get_role(request))
    return {"success": True, "data": data}

File: web-admin/src/api/smartbi/priceAnomaly.ts
/**
 * Wave2 价格异常威慑引擎 — Python gold API client.
 *
 * Wraps /api/smartbi/gold/price-anomaly/* (Python backend). pythonFetch
 * auto-camelCases responses (transformKeys), so the interfaces below describe
 * the camelCase shape the caller receives — including the {success, data,
 * message} envelope these endpoints emit (unlike the bare gold-reads endpoints).
 *
 * 邓总差异化护城河: 检测同类物料相邻采购单价异常 (洗洁精 110→150、青菜 2.5→2),
 * 要求供应商录入解释 (威慑非处罚), 连续异常累计标高风险供应商。
 */
import { pythonFetch, PYTHON_LLM_TIMEOUT_MS } from './common';

/** 标准异常原因码 (与后端 VALID_REASON_CODES 对齐, Rule 3: dropdown 约束)。 */
export type ReasonCode = 'SEASONAL' | 'MARKET_RISE' | 'SPEC_CHANGE' | 'OTHER';

export const REASON_OPTIONS: Array<{ value: ReasonCode; label: string }> = [
  { value: 'SEASONAL', label: '季节性' },
  { value: 'MARKET_RISE', label: '市场涨价' },
  { value: 'SPEC_CHANGE', label: '规格变化' },
  { value: 'OTHER', label: '其他' },
];

/** 一条检测到的价格异常。 */
export interface PriceAnomaly {
  normalizedName: string;
  ingredientName: string | null;
  supplierId: string | null;
  supplierName: string | null;
  unit: string | null;
  anomalyDeliveryDate: string | null; // YYYY-MM-DD
  oldPrice: number | null; // 立即前一次价 (RBAC: 非 price-view 角色为 null)
  newPrice: number | null; // 最近一次异常价 (RBAC 同上)
  trailingAvg: number | null; // trailing-N 均价 (RBAC 同上)
  deltaPct: number | null; // 偏离率 (率, 始终可见 — 传达威慑信号)
  direction: 'UP' | 'DOWN';
  consecutiveAnomalyCount: number; // 连续同向异常累计 (含本次)
  riskLevel: 'MEDIUM' | 'HIGH'; // 连续三次 → HIGH
}

/** 一条已确认/解释的异常记录。 */
export interface PriceAnomalyAck {
  id: number;
  normalizedName: string;
  ingredientName: string | null;
  supplierId: string | null;
  supplierName: string | null;
  anomalyDeliveryDate: string | null;
  oldPrice: number | null;
  newPrice: number | null;
  deltaPct: number | null;
  reasonCode: ReasonCode;
  reasonNote: string | null;
  acknowledgedBy: string | null;
  createdAt: string | null;
}

interface Envelope<T> {
  success: boolean;
  data: T;
  message?: string;
}

export interface DetectArgs {
  trailingN?: number; // trailing 均价窗口 (default 3)
  epsilonPct?: number; // 异常容限 (default 5%)
  baselineMode?: 'count' | 'days'; // days = 邓总锁定的自身 90 天移动均价
  windowDays?: number; // days 模式窗口天数
}

/** 检测同类物料相邻采购单价异常。 */
export async function detectPriceAnomalies(args: DetectArgs = {}): Promise<PriceAnomaly[]> {
  const p = new URLSearchParams();
  if (args.trailingN !== undefined) p.set('trailing_n', String(args.trailingN));
  if (args.epsilonPct !== undefined) p.set('epsilon_pct', String(args.epsilonPct));
  if (args.baselineMode !== undefined) p.set('baseline_mode', args.baselineMode);
  if (args.windowDays !== undefined) p.set('window_days', String(args.windowDays));
  const qs = p.toString();
  const resp = (await pythonFetch(
    `/api/smartbi/gold/price-anomaly/detect${qs ? `?${qs}` : ''}`,
    { timeoutMs: PYTHON_LLM_TIMEOUT_MS },
  )) as Envelope<PriceAnomaly[]>;
  if (!resp.success) throw new Error(resp.message || '价格异常检测失败');
  return resp.data || [];
}

export interface AckBody {
  normalizedName: string;
  ingredientName?: string | null;
  supplierId?: string | null;
  supplierName?: string | null;
  anomalyDeliveryDate: string; // YYYY-MM-DD
  oldPrice?: number | null;
  newPrice: number;
  deltaPct?: number | null;
  reasonCode: ReasonCode;
  reasonNote?: string | null;
}

/** 录入供应商解释/确认 (威慑非处罚, 幂等)。 */
export async function ackPriceAnomaly(body: AckBody): Promise<{ id: number }> {
  const resp = (await pythonFetch('/api/smartbi/gold/price-anomaly/ack', {
    method: 'POST',
    body: JSON.stringify(body),
    timeoutMs: PYTHON_LLM_TIMEOUT_MS,
  })) as Envelope<{ id: number }>;
  if (!resp.success) throw new Error(resp.message || '解释录入失败');
  return resp.data;
}

/** 已确认异常列表 (报警去重 + 留痕)。 */
export async function listPriceAnomalyAcks(ingredient?: string): Promise<PriceAnomalyAck[]> {
  const qs = ingredient ? `?ingredient=${encodeURIComponent(ingredient)}` : '';
  const resp = (await pythonFetch(`/api/smartbi/gold/price-anomaly/acks${qs}`, {
    timeoutMs: PYTHON_LLM_TIMEOUT_MS,
  })) as Envelope<PriceAnomalyAck[]>;
  if (!resp.success) throw new Error(resp.message || '已确认列表加载失败');
  return resp.data || [];
}

/** 一个进价趋势点 (供应商采购单价随时间)。 */
export interface SupplierPricePoint {
  deliveryDate: string | null; // YYYY-MM-DD
  unitPrice: number | null; // RBAC: 非 price-view 角色为 null
  supplierName: string | null;
  quantity: number | null;
}

/**
 * 单个食材的进价趋势 — [{deliveryDate, unitPrice, supplierName}] ASC by date。
 * 用于价格异常页的采购单价趋势曲线 (把异常点放回时间序列看得见涨跌)。
 */
export async function getSupplierPriceTrend(ingredient: string, days = 90): Promise<SupplierPricePoint[]> {
  const qs = `?ingredient=${encodeURIComponent(ingredient)}&days=${days}`;
  const resp = (await pythonFetch(`/api/smartbi/gold/supplier-price/trend${qs}`, {
    timeoutMs: PYTHON_LLM_TIMEOUT_MS,
  })) as Envelope<SupplierPricePoint[]>;
  if (!resp.success) throw new Error(resp.message || '进价趋势加载失败');
  return resp.data || [];
}
